File: app/src/main/java/com/jobtick/activities/ChangePasswordActivity.java
package com.jobtick.activities;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import com.jobtick.R;
import android.annotation.SuppressLint;
import com.jobtick.utils.Constant;
import com.jobtick.utils.HttpStatus;
import com.jobtick.widget.ExtendedEntryText;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import timber.log.Timber;


public class ChangePasswordActivity extends ActivityBase implements TextWatcher {

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.toolbar)
    MaterialToolbar toolbar;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.old_password)
    ExtendedEntryText oldPassword;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.new_password)
    ExtendedEntryText newPassword;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.btn_change_password)
    MaterialButton btnChangePassword;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_change_password);
        ButterKnife.bind(this);

        initToolbar();
        btnChangePassword.setEnabled(false);
        oldPassword.addTextChangedListener(this);
        newPassword.addTextChangedListener(this);
    }


    private void initToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_back);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("Change Password");
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }


    @OnClick({R.id.btn_change_password})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.btn_change_password:
                if (validation())
                    changePassword();
                break;
        }
    }

    private boolean validation() {
        if (TextUtils.isEmpty(oldPassword.getText().trim())) {
            oldPassword.setError("Enter your old password");
            return false;
        } else if (TextUtils.isEmpty(newPassword.getText().trim())) {
            newPassword.setError("Enter new password");
            return false;
        }
        return true;
    }

    private void changePassword() {
        showProgressDialog();
        String str_old_password = oldPassword.getText().trim();
        String str_new_password = newPassword.getText().trim();

        StringRequest stringRequest = new StringRequest(Request.Method.POST, Constant.URL_CHANGE_PASSWORD,
                response -> {
                    Timber.e(response);
                    hideProgressDialog();
                    try {

                        JSONObject jsonObject = new JSONObject(response);
                        if (jsonObject.has("success") && !jsonObject.isNull("success")) {
                            if (jsonObject.getBoolean("success")) {
                                showSuccessToast("Password Changed Successfully !", this);
                            }
                        }

                    } catch (JSONException e) {
                        Timber.e(String.valueOf(e));
                        e.printStackTrace();

                    }


                },
                new com.android.volley.Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        NetworkResponse networkResponse = error.networkResponse;
                        if (networkResponse != null && networkResponse.data != null) {
                            String jsonError = new String(networkResponse.data);
                            // Print Error!
                            Timber.e(jsonError);
                            if (networkResponse.statusCode == HttpStatus.AUTH_FAILED) {
                                unauthorizedUser();
                                hideProgressDialog();
                                return;
                            }
                            try {
                                JSONObject jsonObject = new JSONObject(jsonError);
                                JSONObject jsonObject_error = jsonObject.getJSONObject("error");

                                if (jsonObject_error.has("message")) {
                                    showToast(jsonObject_error.getString("message"), ChangePasswordActivity.this);
                                }
                                if (jsonObject_error.has("errors")) {
                                    JSONObject jsonObject_errors = jsonObject_error.getJSONObject("errors");

                                    if (jsonObject_errors.has("old_password")) {
                                        JSONArray jsonArray_mobile = jsonObject_errors.getJSONArray("old_password");
                                        String old_password = jsonArray_mobile.getString(0);
                                        oldPassword.setError(old_password);
                                    }
                                    if (jsonObject_errors.has("new_password")) {
                                        JSONArray jsonArray_mobile = jsonObject_errors.getJSONArray("new_password");
                                        String new_password = jsonArray_mobile.getString(0);
                                        newPassword.setError(new_password);
                                    }


                                }

                                //  ((CredentialActivity)requireActivity()).showToast(message,requireActivity());


                            } catch (JSONException e) {
                                e.printStackTrace();
                            }
                        } else {
                            showToast("Something Went Wrong", ChangePasswordActivity.this);
                        }
                        Timber.e(error.toString());
                        hideProgressDialog();
                    }
                }) {


            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<>();

                map1.put("authorization", sessionManager.getTokenType() + " " + sessionManager.getAccessToken());
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                // map1.put("X-Requested-With", "XMLHttpRequest");
                return map1;
            }

            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<>();
                map1.put("old_password", str_old_password);
                map1.put("new_password", str_new_password);
                Timber.e(String.valueOf(map1.size()));
                Timber.e(map1.toString());
                return map1;
            }
        };
        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(ChangePasswordActivity.this);
        requestQueue.add(stringRequest);
    }


    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable editable) {
        boolean enabled = oldPassword.getText().length() > 0 && newPassword.getText().length() > 0;
        btnChangePassword.setEnabled(enabled);
    }
}

File: app/src/main/java/com/jobtick/android/activities/ChangePasswordActivity.kt
package com.jobtick.android.activities

import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.MenuItem
import androidx.lifecycle.ViewModelProvider
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.button.MaterialButton
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.HttpStatus
import com.jobtick.android.viewmodel.ChangePassViewModel
import com.jobtick.android.widget.ExtendedEntryText
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.util.*

class ChangePasswordActivity : ActivityBase(), TextWatcher {
    private var toolbar: MaterialToolbar? = null
    private var oldPassword: ExtendedEntryText? = null
    private var newPassword: ExtendedEntryText? = null
    private var btnChangePassword: MaterialButton? = null
    private lateinit var changePassViewModel: ChangePassViewModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_change_password)
        initIDS()
        initToolbar()
        initVM()
        btnChangePassword!!.isEnabled = false
        oldPassword!!.addTextChangedListener(this)
        newPassword!!.addTextChangedListener(this)
    }

    private fun initIDS() {
        toolbar = findViewById(R.id.toolbar)
        oldPassword = findViewById(R.id.old_password)
        newPassword = findViewById(R.id.new_password)
        btnChangePassword = findViewById(R.id.btn_change_password)
        btnChangePassword!!.setOnClickListener {
            if (validation()) changePassword()
        }
    }

    private fun initToolbar() {
        toolbar!!.setNavigationIcon(R.drawable.ic_back)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.title = "Change Password"
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }


    private fun validation(): Boolean {
        return when {
            TextUtils.isEmpty(oldPassword!!.text.trim { it <= ' ' }) -> {
                oldPassword!!.setError("Enter your old password")
                false
            }
            TextUtils.isEmpty(newPassword!!.text.trim { it <= ' ' }) -> {
                newPassword!!.setError("Enter new password")
                false
            }
            oldPassword!!.text.length < 8 -> {
                oldPassword!!.setError("Password must be 8 character or more")
                false
            }
            else -> true
        }
    }

    private fun changePassword() {
        showProgressDialog()
        val strOldPassword = oldPassword!!.text.trim { it <= ' ' }
        val strNewPassword = newPassword!!.text.trim { it <= ' ' }
        changePassViewModel.changePass(sessionManager!!.accessToken, Volley.newRequestQueue(this), strOldPassword, strNewPassword)
    }

    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun afterTextChanged(editable: Editable) {
        val enabled = oldPassword!!.text.isNotEmpty() && newPassword!!.text.isNotEmpty()
        btnChangePassword!!.isEnabled = enabled
    }

    private fun initVM() {
        changePassViewModel = ViewModelProvider(this).get(ChangePassViewModel::class.java)
        changePassViewModel.changePassResponse().observe(this) {
            hideProgressDialog()
            try {

                if (it.has("success") && !it.isNull("success")) {
                    if (it.getBoolean("success")) {
                        showSuccessToast("Password Changed Successfully !", this)
                    }
                }
            } catch (e: JSONException) {
                Timber.e(e.toString())
                e.printStackTrace()
            }
        }
        changePassViewModel.getError2().observe(this, androidx.lifecycle.Observer {
            this.showToast("Something went wrong", this)

        })
        changePassViewModel.getError().observe(this, androidx.lifecycle.Observer { networkResponse ->
            if (networkResponse?.data != null) {
                val jsonError = String(networkResponse.data)
                // Print Error!
                Timber.e(jsonError)
                if (networkResponse.statusCode == HttpStatus.AUTH_FAILED) {
                    unauthorizedUser()
                    hideProgressDialog()
                    return@Observer
                }
                try {
                    val jsonObject = JSONObject(jsonError)
                    val jsonObjectError = jsonObject.getJSONObject("error")
                    if (jsonObjectError.has("message")) {
                        showToast(jsonObjectError.getString("message"), this@ChangePasswordActivity)
                    }
                    if (jsonObjectError.has("errors")) {
                        val jsonObjectErrors = jsonObjectError.getJSONObject("errors")
                        if (jsonObjectErrors.has("old_password")) {
                            val jsonArrayMobile = jsonObjectErrors.getJSONArray("old_password")
                            val oldPassword1 = jsonArrayMobile.getString(0)
                            oldPassword!!.setError(oldPassword1)
                        }
                        if (jsonObjectErrors.has("new_password")) {
                            val jsonArrayMobile = jsonObjectErrors.getJSONArray("new_password")
                            val newPassword1 = jsonArrayMobile.getString(0)
                            newPassword!!.setError(newPassword1)
                        }
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            } else {
                showToast("Something Went Wrong", this@ChangePasswordActivity)
            }
            hideProgressDialog()
        })
    }
}